diff -ur /usr/lib/python2.7/site-packages/cinder.orig/transfer/api.py /usr/lib/python2.7/site-packages/cinder/transfer/api.py
--- /usr/lib/python2.7/site-packages/cinder.orig/transfer/api.py	2016-11-07 14:39:54.854331962 +0100
+++ /usr/lib/python2.7/site-packages/cinder/transfer/api.py	2016-11-07 14:40:17.399029213 +0100
@@ -62,7 +62,7 @@
     def delete(self, context, transfer_id):
         """Make the RPC call to delete a volume transfer."""
         transfer = self.db.transfer_get(context, transfer_id)
-        volume_api.check_policy(context, 'delete_transfer', target=transfer)
+        volume_api.check_policy(context, 'delete_transfer', transfer)
 
         volume_ref = self.db.volume_get(context, transfer.volume_id)
         volume_utils.notify_about_volume_usage(context, volume_ref,
@@ -105,7 +105,7 @@
     def create(self, context, volume_id, display_name):
         """Creates an entry in the transfers table."""
         volume_ref = self.db.volume_get(context, volume_id)
-        volume_api.check_policy(context, 'create_transfer', target=volume_ref)
+        volume_api.check_policy(context, 'create_transfer', volume_ref)
         LOG.info(_LI("Generating transfer record for volume %s"), volume_id)
         if volume_ref['status'] != "available":
             raise exception.InvalidVolume(reason=_("status must be available"))
