diff -ru cinder.orig/transfer/api.py cinder/transfer/api.py
--- cinder.orig/transfer/api.py 2016-05-26 15:34:09.144336470 +0200
+++ cinder/transfer/api.py      2016-05-26 16:06:47.652323724 +0200
@@ -61,8 +61,8 @@
 
     def delete(self, context, transfer_id):
         """Make the RPC call to delete a volume transfer."""
-        volume_api.check_policy(context, 'delete_transfer')
         transfer = self.db.transfer_get(context, transfer_id)
+        volume_api.check_policy(context, 'delete_transfer', target=transfer)
 
         volume_ref = self.db.volume_get(context, transfer.volume_id)
         volume_utils.notify_about_volume_usage(context, volume_ref,
@@ -104,9 +104,9 @@
 
     def create(self, context, volume_id, display_name):
         """Creates an entry in the transfers table."""
-        volume_api.check_policy(context, 'create_transfer')
-        LOG.info(_LI("Generating transfer record for volume %s"), volume_id)
         volume_ref = self.db.volume_get(context, volume_id)
+        volume_api.check_policy(context, 'create_transfer', target=volume_ref)
+        LOG.info(_LI("Generating transfer record for volume %s"), volume_id)
         if volume_ref['status'] != "available":
             raise exception.InvalidVolume(reason=_("status must be available"))
 
diff -ru cinder.orig/volume/api.py cinder/volume/api.py
--- cinder.orig/volume/api.py   2016-05-26 15:38:27.305292600 +0200
+++ cinder/volume/api.py        2016-05-26 15:09:15.769295413 +0200
@@ -16,7 +16,7 @@
 
 """Handles all requests relating to volumes."""
 
-import ast
+
 import collections
 import datetime
 import functools
@@ -206,7 +206,10 @@
                source_replica=None, consistencygroup=None,
                cgsnapshot=None, multiattach=False, source_cg=None):
 
-        check_policy(context, 'create')
+        if snapshot:
+            check_policy(context, 'update_snapshot', snapshot.volume_id)
+        else:
+            check_policy(context, 'create')
 
         # NOTE(jdg): we can have a create without size if we're
         # doing a create from snap or volume.  Currently
@@ -258,7 +261,6 @@
             msg = _("No volume_type should be provided when creating test "
                     "replica.")
             raise exception.InvalidInput(reason=msg)
-
         if snapshot and volume_type:
             if volume_type['id'] != snapshot.volume_type_id:
                 if not self._retype_is_possible(context,
@@ -505,7 +507,6 @@
         return volumes
 
     def get_snapshot(self, context, snapshot_id):
-        check_policy(context, 'get_snapshot')
         snapshot = objects.Snapshot.get_by_id(context, snapshot_id)
 
         # FIXME(jdg): The objects don't have the db name entries
@@ -926,9 +927,11 @@
                  resource=result)
         return result
 
-    @wrap_check_policy
     def delete_snapshot(self, context, snapshot, force=False,
                         unmanage_only=False):
+        rv = self.db.volume_get(context, snapshot.volume_id)
+        volume = dict(rv)
+        check_policy(context, 'delete_snapshot', volume)
         if not force and snapshot.status not in ["available", "error"]:
             LOG.error(_LE('Unable to delete snapshot: %(snap_id)s, '
                           'due to invalid status. '
@@ -956,8 +959,11 @@
         LOG.info(_LI("Snapshot delete request issued successfully."),
                  resource=snapshot)
 
-    @wrap_check_policy
     def update_snapshot(self, context, snapshot, fields):
+        rv = self.db.volume_get(context, snapshot.volume_id)
+        volume = dict(rv)
+        LOG.error("\nXXXXXX %s" % volume)
+        check_policy(context, 'update_snapshot', volume)
         snapshot.update(fields)
         snapshot.save()
 

