diff -ur /usr/lib/python2.7/dist-packages/cinder.orig/transfer/api.py /usr/lib/python2.7/dist-packages/cinder/transfer/api.py
--- /usr/lib/python2.7/dist-packages/cinder.orig/transfer/api.py	2018-01-16 20:29:21.870989403 +0100
+++ /usr/lib/python2.7/dist-packages/cinder/transfer/api.py	2018-01-16 20:31:10.454278041 +0100
@@ -62,8 +62,8 @@
 
     def delete(self, context, transfer_id):
         """Make the RPC call to delete a volume transfer."""
-        volume_api.check_policy(context, 'delete_transfer')
         transfer = self.db.transfer_get(context, transfer_id)
+        volume_api.check_policy(context, 'delete_transfer', target=transfer)
 
         volume_ref = self.db.volume_get(context, transfer.volume_id)
         volume_utils.notify_about_volume_usage(context, volume_ref,
@@ -111,9 +111,9 @@
 
     def create(self, context, volume_id, display_name):
         """Creates an entry in the transfers table."""
-        volume_api.check_policy(context, 'create_transfer')
-        LOG.info(_LI("Generating transfer record for volume %s"), volume_id)
         volume_ref = self.db.volume_get(context, volume_id)
+        volume_api.check_policy(context, 'create_transfer', target=volume_ref)
+        LOG.info(_LI("Generating transfer record for volume %s"), volume_id)
         if volume_ref['status'] != "available":
             raise exception.InvalidVolume(reason=_("status must be available"))
 
diff -ur /usr/lib/python2.7/dist-packages/cinder.orig/volume/api.py /usr/lib/python2.7/dist-packages/cinder/volume/api.py
--- /usr/lib/python2.7/dist-packages/cinder.orig/volume/api.py	2018-01-16 20:29:21.914989115 +0100
+++ /usr/lib/python2.7/dist-packages/cinder/volume/api.py	2018-01-16 20:33:45.445262639 +0100
@@ -213,7 +213,11 @@
                source_replica=None, consistencygroup=None,
                cgsnapshot=None, multiattach=False, source_cg=None):
 
-        check_policy(context, 'create')
+        if snapshot:
+            check_policy(context, 'update_snapshot', snapshot.volume_id)
+        else:
+            check_policy(context, 'create')
 
         # NOTE(jdg): we can have a create without size if we're
         # doing a create from snap or volume.  Currently
@@ -524,6 +528,10 @@
 
     def get_snapshot(self, context, snapshot_id):
         snapshot = objects.Snapshot.get_by_id(context, snapshot_id)
+        rv = self.db.volume_get(context, snapshot.volume_id)
+        volume = dict(rv)
+        check_policy(context, 'get_snapshot', volume)
+
 
         # FIXME(jdg): The objects don't have the db name entries
         # so build the resource tag manually for now.
@@ -940,9 +948,11 @@
                  resource=result)
         return result
 
-    @wrap_check_policy
     def delete_snapshot(self, context, snapshot, force=False,
                         unmanage_only=False):
+        rv = self.db.volume_get(context, snapshot.volume_id)
+        volume = dict(rv)
+        check_policy(context, 'delete_snapshot', volume)
         # Build required conditions for conditional update
         expected = {'cgsnapshot_id': None}
         # If not force deleting we have status conditions
@@ -965,8 +975,10 @@
         LOG.info(_LI("Snapshot delete request issued successfully."),
                  resource=snapshot)
 
-    @wrap_check_policy
     def update_snapshot(self, context, snapshot, fields):
+        rv = self.db.volume_get(context, snapshot.volume_id)
+        volume = dict(rv)
+        check_policy(context, 'update_snapshot', volume)
         snapshot.update(fields)
         snapshot.save()
 
