diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/admin_password.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/admin_password.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/admin_password.py	2016-09-19 18:19:25.747238632 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/admin_password.py	2016-09-19 18:19:47.509242199 +0200
@@ -43,10 +43,10 @@
     @validation.schema(admin_password.change_password)
     def change_password(self, req, id, body):
         context = req.environ['nova.context']
-        authorize(context)
+        instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
 
         password = body['changePassword']['adminPass']
-        instance = common.get_instance(self.compute_api, context, id)
         try:
             self.compute_api.set_admin_password(context, instance, password)
         except exception.InstanceUnknownCell as e:
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/console_output.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/console_output.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/console_output.py	2016-09-19 18:19:25.759238634 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/console_output.py	2016-09-19 18:19:47.509242199 +0200
@@ -41,9 +41,9 @@
     def get_console_output(self, req, id, body):
         """Get text console output."""
         context = req.environ['nova.context']
-        authorize(context)
-
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
+
         length = body['os-getConsoleOutput'].get('length')
         # TODO(cyeoh): In a future API update accept a length of -1
         # as meaning unlimited length (convert to None)
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/lock_server.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/lock_server.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/lock_server.py	2016-09-19 18:19:25.761238635 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/lock_server.py	2016-09-19 18:19:47.509242199 +0200
@@ -34,8 +34,8 @@
     def _lock(self, req, id, body):
         """Lock a server instance."""
         context = req.environ['nova.context']
-        authorize(context, action='lock')
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='lock', target=instance)
         self.compute_api.lock(context, instance)
 
     @wsgi.response(202)
@@ -44,8 +44,8 @@
     def _unlock(self, req, id, body):
         """Unlock a server instance."""
         context = req.environ['nova.context']
-        authorize(context, action='unlock')
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='unlock', target=instance)
         if not self.compute_api.is_expected_locked_by(context, instance):
             authorize(context, target=instance,
                       action='unlock:unlock_override')
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/pause_server.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/pause_server.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/pause_server.py	2016-09-19 18:19:25.764238635 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/pause_server.py	2016-09-19 18:19:47.509242199 +0200
@@ -37,8 +37,8 @@
     def _pause(self, req, id, body):
         """Permit Admins to pause the server."""
         ctxt = req.environ['nova.context']
-        authorize(ctxt, action='pause')
         server = common.get_instance(self.compute_api, ctxt, id)
+        authorize(ctxt, action='pause', target=server)
         try:
             self.compute_api.pause(ctxt, server)
         except exception.InstanceIsLocked as e:
@@ -58,8 +58,8 @@
     def _unpause(self, req, id, body):
         """Permit Admins to unpause the server."""
         ctxt = req.environ['nova.context']
-        authorize(ctxt, action='unpause')
         server = common.get_instance(self.compute_api, ctxt, id)
+        authorize(ctxt, action='unpause', target=server)
         try:
             self.compute_api.unpause(ctxt, server)
         except exception.InstanceIsLocked as e:
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/remote_consoles.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/remote_consoles.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/remote_consoles.py	2016-09-19 18:19:26.013238676 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/remote_consoles.py	2016-09-19 18:19:47.509242199 +0200
@@ -44,13 +44,13 @@
     def get_vnc_console(self, req, id, body):
         """Get text console output."""
         context = req.environ['nova.context']
-        authorize(context)
+        instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
 
         # If type is not supplied or unknown, get_vnc_console below will cope
         console_type = body['os-getVNCConsole'].get('type')
 
         try:
-            instance = common.get_instance(self.compute_api, context, id)
             output = self.compute_api.get_vnc_console(context,
                                                       instance,
                                                       console_type)
@@ -73,13 +73,13 @@
     def get_spice_console(self, req, id, body):
         """Get text console output."""
         context = req.environ['nova.context']
-        authorize(context)
+        instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
 
         # If type is not supplied or unknown, get_spice_console below will cope
         console_type = body['os-getSPICEConsole'].get('type')
 
         try:
-            instance = common.get_instance(self.compute_api, context, id)
             output = self.compute_api.get_spice_console(context,
                                                         instance,
                                                         console_type)
@@ -102,12 +102,12 @@
     def get_rdp_console(self, req, id, body):
         """Get text console output."""
         context = req.environ['nova.context']
-        authorize(context)
+        instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
 
         # If type is not supplied or unknown, get_rdp_console below will cope
         console_type = body['os-getRDPConsole'].get('type')
 
-        instance = common.get_instance(self.compute_api, context, id)
         try:
             # NOTE(mikal): get_rdp_console() can raise InstanceNotFound, so
             # we still need to catch it here.
@@ -133,12 +133,12 @@
     def get_serial_console(self, req, id, body):
         """Get connection to a serial console."""
         context = req.environ['nova.context']
-        authorize(context)
+        instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, target=instance)
 
         # If type is not supplied or unknown get_serial_console below will cope
         console_type = body['os-getSerialConsole'].get('type')
         try:
-            instance = common.get_instance(self.compute_api, context, id)
             output = self.compute_api.get_serial_console(context,
                                                          instance,
                                                          console_type)
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/server_metadata.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/server_metadata.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/server_metadata.py	2016-09-19 18:19:25.755238634 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/server_metadata.py	2016-09-19 18:19:47.510242200 +0200
@@ -55,7 +55,8 @@
     def index(self, req, server_id):
         """Returns the list of metadata for a given instance."""
         context = req.environ['nova.context']
-        authorize(context, action='index')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='index', target=server)
         return {'metadata': self._get_metadata(context, server_id)}
 
     @extensions.expected_errors((400, 403, 404, 409, 413))
@@ -65,7 +66,8 @@
     def create(self, req, server_id, body):
         metadata = body['metadata']
         context = req.environ['nova.context']
-        authorize(context, action='create')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='create', target=server)
         new_metadata = self._update_instance_metadata(context,
                                                       server_id,
                                                       metadata,
@@ -77,7 +79,8 @@
     @validation.schema(server_metadata.update)
     def update(self, req, server_id, id, body):
         context = req.environ['nova.context']
-        authorize(context, action='update')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='update', target=server)
         meta_item = body['meta']
         if id not in meta_item:
             expl = _('Request body and URI mismatch')
@@ -94,7 +97,8 @@
     @validation.schema(server_metadata.update_all)
     def update_all(self, req, server_id, body):
         context = req.environ['nova.context']
-        authorize(context, action='update_all')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='update_all', target=server)
         metadata = body['metadata']
         new_metadata = self._update_instance_metadata(context,
                                                       server_id,
@@ -129,7 +133,8 @@
     def show(self, req, server_id, id):
         """Return a single metadata item."""
         context = req.environ['nova.context']
-        authorize(context, action='show')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='show', target=server)
         data = self._get_metadata(context, server_id)
 
         try:
@@ -143,14 +148,14 @@
     def delete(self, req, server_id, id):
         """Deletes an existing metadata."""
         context = req.environ['nova.context']
-        authorize(context, action='delete')
+        server = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, action='delete', target=server)
         metadata = self._get_metadata(context, server_id)
 
         if id not in metadata:
             msg = _("Metadata item was not found")
             raise exc.HTTPNotFound(explanation=msg)
 
-        server = common.get_instance(self.compute_api, context, server_id)
         try:
             self.compute_api.delete_instance_metadata(context, server, id)
 
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/server_password.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/server_password.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/server_password.py	2016-09-19 18:19:25.755238634 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/server_password.py	2016-09-19 18:19:47.510242200 +0200
@@ -34,8 +34,8 @@
     @extensions.expected_errors(404)
     def index(self, req, server_id):
         context = req.environ['nova.context']
-        authorize(context)
         instance = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, target=instance)
 
         passw = password.extract_password(instance)
         return {'password': passw or ''}
@@ -50,8 +50,8 @@
         """
 
         context = req.environ['nova.context']
-        authorize(context)
         instance = common.get_instance(self.compute_api, context, server_id)
+        authorize(context, target=instance)
         meta = password.convert_password(context, None)
         instance.system_metadata.update(meta)
         instance.save()
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/servers.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/servers.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/servers.py	2016-09-19 18:19:25.758238634 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/servers.py	2016-09-19 18:19:47.510242200 +0200
@@ -753,8 +753,8 @@
         resize_schema['properties']['resize']['properties'].update(schema)
 
     def _delete(self, context, req, instance_uuid):
-        authorize(context, action='delete')
         instance = self._get_server(context, req, instance_uuid)
+        authorize(context, action='delete', target=instance)
         if CONF.reclaim_instance_interval:
             try:
                 self.compute_api.soft_delete(context, instance)
@@ -774,7 +774,8 @@
 
         ctxt = req.environ['nova.context']
         update_dict = {}
-        authorize(ctxt, action='update')
+        instance = self._get_server(ctxt, req, id, is_detail=True)
+        authorize(ctxt, action='update', target=instance)
 
         if 'name' in body['server']:
             update_dict['display_name'] = common.normalize_name(
@@ -784,7 +785,6 @@
             self.update_extension_manager.map(self._update_extension_point,
                                               body['server'], update_dict)
 
-        instance = self._get_server(ctxt, req, id, is_detail=True)
         try:
             # NOTE(mikal): this try block needs to stay because save() still
             # might throw an exception.
@@ -804,8 +804,9 @@
     @wsgi.action('confirmResize')
     def _action_confirm_resize(self, req, id, body):
         context = req.environ['nova.context']
-        authorize(context, action='confirm_resize')
         instance = self._get_server(context, req, id)
+        authorize(context, action='confirm_resize', target=instance)
+
         try:
             self.compute_api.confirm_resize(context, instance)
         except exception.InstanceUnknownCell as e:
@@ -824,8 +825,9 @@
     @wsgi.action('revertResize')
     def _action_revert_resize(self, req, id, body):
         context = req.environ['nova.context']
-        authorize(context, action='revert_resize')
         instance = self._get_server(context, req, id)
+        authorize(context, action='revert_resize', target=instance)
+
         try:
             self.compute_api.revert_resize(context, instance)
         except exception.InstanceUnknownCell as e:
@@ -850,8 +852,8 @@
 
         reboot_type = body['reboot']['type'].upper()
         context = req.environ['nova.context']
-        authorize(context, action='reboot')
         instance = self._get_server(context, req, id)
+        authorize(context, action='reboot', target=instance)
 
         try:
             self.compute_api.reboot(context, instance, reboot_type)
@@ -864,8 +866,8 @@
     def _resize(self, req, instance_id, flavor_id, **kwargs):
         """Begin the resize process with given instance/flavor."""
         context = req.environ["nova.context"]
-        authorize(context, action='resize')
         instance = self._get_server(context, req, instance_id)
+        authorize(context, action='resize', target=instance)
 
         try:
             self.compute_api.resize(context, instance, flavor_id, **kwargs)
@@ -959,6 +961,10 @@
     @validation.schema(schema_server_resize)
     def _action_resize(self, req, id, body):
         """Resizes a given instance to the flavor size requested."""
+        context = req.environ['nova.context']
+        instance = self._get_server(context, req, id)
+        authorize(context, action='resize', target=instance)
+
         resize_dict = body['resize']
         flavor_ref = str(resize_dict["flavorRef"])
 
@@ -985,8 +991,8 @@
         password = self._get_server_admin_password(rebuild_dict)
 
         context = req.environ['nova.context']
-        authorize(context, action='rebuild')
         instance = self._get_server(context, req, id)
+        authorize(context, action='rebuild', target=instance)
 
         attr_map = {
             'name': 'display_name',
@@ -1059,7 +1065,8 @@
     def _action_create_image(self, req, id, body):
         """Snapshot a server instance."""
         context = req.environ['nova.context']
-        authorize(context, action='create_image')
+        instance = self._get_server(context, req, id)
+        authorize(context, action='create_image', target=instance)
 
         entity = body["createImage"]
         image_name = common.normalize_name(entity["name"])
@@ -1067,8 +1074,6 @@
 
         common.check_img_metadata_properties_quota(context, metadata)
 
-        instance = self._get_server(context, req, id)
-
         bdms = objects.BlockDeviceMappingList.get_by_instance_uuid(
                     context, instance.uuid)
 
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/shelve.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/shelve.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/shelve.py	2016-09-19 18:19:25.759238634 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/shelve.py	2016-09-19 18:19:47.510242200 +0200
@@ -38,9 +38,9 @@
     def _shelve(self, req, id, body):
         """Move an instance into shelved mode."""
         context = req.environ["nova.context"]
-        authorize(context, action='shelve')
-
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='shelve', target=instance)
+
         try:
             self.compute_api.shelve(context, instance)
         except exception.InstanceUnknownCell as e:
@@ -57,9 +57,9 @@
     def _shelve_offload(self, req, id, body):
         """Force removal of a shelved instance from the compute node."""
         context = req.environ["nova.context"]
-        authorize(context, action='shelve_offload')
-
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='shelve_offload', target=instance)
+
         try:
             self.compute_api.shelve_offload(context, instance)
         except exception.InstanceUnknownCell as e:
@@ -77,8 +77,9 @@
     def _unshelve(self, req, id, body):
         """Restore an instance from shelved mode."""
         context = req.environ["nova.context"]
-        authorize(context, action='unshelve')
         instance = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='unshelve', target=instance)
+
         try:
             self.compute_api.unshelve(context, instance)
         except exception.InstanceUnknownCell as e:
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/suspend_server.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/suspend_server.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/suspend_server.py	2016-09-19 18:19:25.760238635 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/suspend_server.py	2016-09-19 18:19:47.511242200 +0200
@@ -37,9 +37,9 @@
     def _suspend(self, req, id, body):
         """Permit admins to suspend the server."""
         context = req.environ['nova.context']
-        authorize(context, action='suspend')
+        server = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='suspend', target=server)
         try:
-            server = common.get_instance(self.compute_api, context, id)
             self.compute_api.suspend(context, server)
         except exception.InstanceUnknownCell as e:
             raise exc.HTTPNotFound(explanation=e.format_message())
@@ -55,9 +55,9 @@
     def _resume(self, req, id, body):
         """Permit admins to resume the server from suspend."""
         context = req.environ['nova.context']
-        authorize(context, action='resume')
+        server = common.get_instance(self.compute_api, context, id)
+        authorize(context, action='resume', target=server)
         try:
-            server = common.get_instance(self.compute_api, context, id)
             self.compute_api.resume(context, server)
         except exception.InstanceUnknownCell as e:
             raise exc.HTTPNotFound(explanation=e.format_message())
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/volumes.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/volumes.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/compute/volumes.py	2016-09-19 18:19:25.764238635 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/volumes.py	2016-09-19 18:19:47.511242200 +0200
@@ -511,7 +511,7 @@
         res = [entity_maker(context, snapshot) for snapshot in limited_list]
         return {'snapshots': res}
 
-    @extensions.expected_errors(400)
+    @extensions.expected_errors((400, 403))
     @validation.schema(volumes_schema.snapshot_create)
     def create(self, req, body):
         """Creates a new snapshot."""
@@ -528,9 +528,12 @@
         else:
             create_func = self.volume_api.create_snapshot
 
-        new_snapshot = create_func(context, volume_id,
-                                   snapshot.get('display_name'),
-                                   snapshot.get('display_description'))
+        try:
+            new_snapshot = create_func(context, volume_id,
+                                       snapshot.get('display_name'),
+                                       snapshot.get('display_description'))
+        except exception.OverQuota as e:
+            raise exc.HTTPForbidden(explanation=e.format_message())
 
         retval = _translate_snapshot_detail_view(context, new_snapshot)
         return {'snapshot': retval}
diff -ur /usr/lib/python2.7/site-packages/nova.orig/api/openstack/extensions.py /usr/lib/python2.7/site-packages/nova/api/openstack/extensions.py
--- /usr/lib/python2.7/site-packages/nova.orig/api/openstack/extensions.py	2016-09-19 18:19:26.044238681 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/extensions.py	2016-09-19 18:19:47.508242199 +0200
@@ -484,9 +484,9 @@
                         t_errors = errors
                     if exc.code in t_errors:
                         raise
-                elif isinstance(exc, exception.PolicyNotAuthorized):
+                elif isinstance(exc, exception.Forbidden):
                     # Note(cyeoh): Special case to handle
-                    # PolicyNotAuthorized exceptions so every
+                    # Forbidden exceptions so every
                     # extension method does not need to wrap authorize
                     # calls. ResourceExceptionHandler silently
                     # converts NotAuthorized to HTTPForbidden
diff -ur /usr/lib/python2.7/site-packages/nova.orig/volume/cinder.py /usr/lib/python2.7/site-packages/nova/volume/cinder.py
--- /usr/lib/python2.7/site-packages/nova.orig/volume/cinder.py	2016-09-19 18:19:26.861238815 +0200
+++ /usr/lib/python2.7/site-packages/nova/volume/cinder.py	2016-09-19 18:19:47.511242200 +0200
@@ -19,6 +19,7 @@
 """
 
 import copy
+import functools
 import sys
 
 from cinderclient import client as cinder_client
@@ -205,6 +206,27 @@
 
     return d
 
+def translate_cinder_exception(method):
+    """Transforms a cinder exception but keeps its traceback intact."""
+    @functools.wraps(method)
+    def wrapper(self, ctx, *args, **kwargs):
+        try:
+            res = method(self, ctx, *args, **kwargs)
+        except (cinder_exception.ConnectionError,
+                keystone_exception.ConnectionError):
+            exc_type, exc_value, exc_trace = sys.exc_info()
+            _reraise(exception.CinderConnectionFailed(
+                reason=six.text_type(exc_value)))
+        except (keystone_exception.BadRequest,
+                cinder_exception.BadRequest):
+            exc_type, exc_value, exc_trace = sys.exc_info()
+            _reraise(exception.InvalidInput(reason=six.text_type(exc_value)))
+        except (keystone_exception.Forbidden,
+                cinder_exception.Forbidden):
+            exc_type, exc_value, exc_trace = sys.exc_info()
+            _reraise(exception.Forbidden(six.text_type(exc_value)))
+        return res
+    return wrapper
 
 def translate_volume_exception(method):
     """Transforms the exception for the volume but keeps its traceback intact.
@@ -212,25 +234,12 @@
     def wrapper(self, ctx, volume_id, *args, **kwargs):
         try:
             res = method(self, ctx, volume_id, *args, **kwargs)
-        except (cinder_exception.ClientException,
-                keystone_exception.ClientException):
-            exc_type, exc_value, exc_trace = sys.exc_info()
-            if isinstance(exc_value, (keystone_exception.NotFound,
-                                      cinder_exception.NotFound)):
-                exc_value = exception.VolumeNotFound(volume_id=volume_id)
-            elif isinstance(exc_value, (keystone_exception.BadRequest,
-                                        cinder_exception.BadRequest)):
-                exc_value = exception.InvalidInput(
-                    reason=six.text_type(exc_value))
-            six.reraise(exc_value, None, exc_trace)
-        except (cinder_exception.ConnectionError,
-                keystone_exception.ConnectionError):
-            exc_type, exc_value, exc_trace = sys.exc_info()
-            exc_value = exception.CinderConnectionFailed(
-                reason=six.text_type(exc_value))
-            six.reraise(exc_value, None, exc_trace)
+        except (keystone_exception.NotFound, cinder_exception.NotFound):
+            _reraise(exception.VolumeNotFound(volume_id=volume_id))
+        except cinder_exception.OverLimit:
+            _reraise(exception.OverQuota(overs='volumes'))
         return res
-    return wrapper
+    return translate_cinder_exception(wrapper)
 
 
 def translate_snapshot_exception(method):
@@ -240,21 +249,29 @@
     def wrapper(self, ctx, snapshot_id, *args, **kwargs):
         try:
             res = method(self, ctx, snapshot_id, *args, **kwargs)
-        except (cinder_exception.ClientException,
-                keystone_exception.ClientException):
-            exc_type, exc_value, exc_trace = sys.exc_info()
-            if isinstance(exc_value, (keystone_exception.NotFound,
-                                      cinder_exception.NotFound)):
-                exc_value = exception.SnapshotNotFound(snapshot_id=snapshot_id)
-            six.reraise(exc_value, None, exc_trace)
-        except (cinder_exception.ConnectionError,
-                keystone_exception.ConnectionError):
-            exc_type, exc_value, exc_trace = sys.exc_info()
-            reason = six.text_type(exc_value)
-            exc_value = exception.CinderConnectionFailed(reason=reason)
-            six.reraise(exc_value, None, exc_trace)
+        except (keystone_exception.NotFound, cinder_exception.NotFound):
+            _reraise(exception.SnapshotNotFound(snapshot_id=snapshot_id))
         return res
-    return wrapper
+    return translate_cinder_exception(wrapper)
+
+
+def translate_mixed_exceptions(method):
+    """Transforms exceptions that can come from both volumes and snapshots."""
+    def wrapper(self, ctx, res_id, *args, **kwargs):
+        try:
+            res = method(self, ctx, res_id, *args, **kwargs)
+        except (keystone_exception.NotFound, cinder_exception.NotFound):
+            _reraise(exception.VolumeNotFound(volume_id=res_id))
+        except cinder_exception.OverLimit:
+            _reraise(exception.OverQuota(overs='snapshots'))
+        return res
+    return translate_cinder_exception(wrapper)
+
+
+def _reraise(desired_exc):
+    exc_type, exc_value, exc_trace = sys.exc_info()
+    exc_value = desired_exc
+    six.reraise(exc_value, None, exc_trace)
 
 
 class API(object):
@@ -265,6 +282,7 @@
         item = cinderclient(context).volumes.get(volume_id)
         return _untranslate_volume_summary_view(context, item)
 
+    @translate_cinder_exception
     def get_all(self, context, search_opts=None):
         search_opts = search_opts or {}
         items = cinderclient(context).volumes.list(detailed=True,
@@ -374,11 +392,13 @@
         return cinderclient(context).volumes.terminate_connection(volume_id,
                                                                   connector)
 
+    @translate_cinder_exception
     def migrate_volume_completion(self, context, old_volume_id, new_volume_id,
                                   error=False):
         return cinderclient(context).volumes.migrate_volume_completion(
             old_volume_id, new_volume_id, error)
 
+    @translate_volume_exception
     def create(self, context, size, name, description, snapshot=None,
                image_id=None, volume_type=None, metadata=None,
                availability_zone=None):
@@ -404,14 +424,8 @@
             kwargs['name'] = name
             kwargs['description'] = description
 
-        try:
-            item = client.volumes.create(size, **kwargs)
-            return _untranslate_volume_summary_view(context, item)
-        except cinder_exception.OverLimit:
-            raise exception.OverQuota(overs='volumes')
-        except (cinder_exception.BadRequest,
-                keystone_exception.BadRequest) as e:
-            raise exception.InvalidInput(reason=e)
+        item = client.volumes.create(size, **kwargs)
+        return _untranslate_volume_summary_view(context, item)
 
     @translate_volume_exception
     def delete(self, context, volume_id):
@@ -426,6 +440,7 @@
         item = cinderclient(context).volume_snapshots.get(snapshot_id)
         return _untranslate_snapshot_summary_view(context, item)
 
+    @translate_cinder_exception
     def get_all_snapshots(self, context):
         items = cinderclient(context).volume_snapshots.list(detailed=True)
         rvals = []
@@ -435,7 +450,7 @@
 
         return rvals
 
-    @translate_volume_exception
+    @translate_mixed_exceptions
     def create_snapshot(self, context, volume_id, name, description):
         item = cinderclient(context).volume_snapshots.create(volume_id,
                                                              False,
@@ -443,7 +458,7 @@
                                                              description)
         return _untranslate_snapshot_summary_view(context, item)
 
-    @translate_volume_exception
+    @translate_mixed_exceptions
     def create_snapshot_force(self, context, volume_id, name, description):
         item = cinderclient(context).volume_snapshots.create(volume_id,
                                                              True,
@@ -456,6 +471,7 @@
     def delete_snapshot(self, context, snapshot_id):
         cinderclient(context).volume_snapshots.delete(snapshot_id)
 
+    @translate_cinder_exception
     def get_volume_encryption_metadata(self, context, volume_id):
         return cinderclient(context).volumes.get_encryption_metadata(volume_id)
