diff -ur /usr/lib/python2.7/site-packages/nova/api/openstack/compute.orig/contrib/admin_actions.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/contrib/admin_actions.py
--- /usr/lib/python2.7/site-packages/nova/api/openstack/compute.orig/contrib/admin_actions.py	2016-06-24 18:04:01.742000000 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/contrib/admin_actions.py	2016-06-24 18:04:23.931000000 +0200
@@ -56,6 +56,8 @@
         server = common.get_instance(self.compute_api, ctxt, id)
         try:
             self.compute_api.pause(ctxt, server)
+        except exception.PolicyNotAuthorized as e:
+            raise webob.exc.HTTPForbidden(explanation=e.format_message())
         except exception.InstanceIsLocked as e:
             raise exc.HTTPConflict(explanation=e.format_message())
         except exception.InstanceInvalidState as state_error:
@@ -78,6 +80,8 @@
         server = common.get_instance(self.compute_api, ctxt, id)
         try:
             self.compute_api.unpause(ctxt, server)
+        except exception.PolicyNotAuthorized as e:
+            raise webob.exc.HTTPForbidden(explanation=e.format_message())
         except exception.InstanceIsLocked as e:
             raise exc.HTTPConflict(explanation=e.format_message())
         except exception.InstanceInvalidState as state_error:
@@ -100,6 +104,8 @@
         server = common.get_instance(self.compute_api, context, id)
         try:
             self.compute_api.suspend(context, server)
+        except exception.PolicyNotAuthorized as e:
+            raise webob.exc.HTTPForbidden(explanation=e.format_message())
         except exception.InstanceIsLocked as e:
             raise exc.HTTPConflict(explanation=e.format_message())
         except exception.InstanceInvalidState as state_error:
@@ -122,6 +128,8 @@
         server = common.get_instance(self.compute_api, context, id)
         try:
             self.compute_api.resume(context, server)
+        except exception.PolicyNotAuthorized as e:
+            raise webob.exc.HTTPForbidden(explanation=e.format_message())
         except exception.InstanceIsLocked as e:
             raise exc.HTTPConflict(explanation=e.format_message())
         except exception.InstanceInvalidState as state_error:
@@ -206,6 +214,8 @@
         instance = common.get_instance(self.compute_api, context, id)
         try:
             self.compute_api.lock(context, instance)
+        except exception.PolicyNotAuthorized as e:
+            raise webob.exc.HTTPForbidden(explanation=e.format_message())
         except exception.InstanceNotFound:
             msg = _("Server not found")
             raise exc.HTTPNotFound(explanation=msg)
diff -ur /usr/lib/python2.7/site-packages/nova/api/openstack/compute.orig/contrib/server_password.py /usr/lib/python2.7/site-packages/nova/api/openstack/compute/contrib/server_password.py
--- /usr/lib/python2.7/site-packages/nova/api/openstack/compute.orig/contrib/server_password.py	2016-06-24 18:04:01.749000000 +0200
+++ /usr/lib/python2.7/site-packages/nova/api/openstack/compute/contrib/server_password.py	2016-06-24 18:06:05.554000000 +0200
@@ -20,6 +20,7 @@
 from nova.api.openstack import extensions
 from nova.api.openstack import wsgi
 from nova import compute
+from nova import policy
 
 
 authorize = extensions.extension_authorizer('compute', 'server_password')
@@ -32,8 +33,8 @@
 
     def index(self, req, server_id):
         context = req.environ['nova.context']
-        authorize(context)
         instance = common.get_instance(self.compute_api, context, server_id)
+        policy.enforce(context, 'compute:server_password', instance)
 
         passw = password.extract_password(instance)
         return {'password': passw or ''}
